### PR TITLE
gh-111789: Use PyDict_GetItemRef()

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
It is a sample. We perhaps will only merge a part of these changes.

* Issue: gh-111789

### CODE REVIEW

File: Modules/_ctypes/_ctypes.c [hunk 1]
```diff
@@ -243,26 +243,19 @@ PyDict_SetItemProxy(PyObject *dict, PyObject *key, PyObject *item)
...
 _PyDict_GetItemProxy(PyObject *dict, PyObject *key, PyObject **presult)
 {
-    PyObject *item = PyDict_GetItemWithError(dict, key);
-    if (item == NULL) {
-        if (PyErr_Occurred()) {
-            return -1;
-        }
+    int rc = PyDict_GetItemRef(dict, key, presult);
+    if (rc <= 0) {  // error or not found
         *presult = NULL;
```
Comment (vstinner) [inline Modules/_ctypes/_ctypes.c]: Is it now redundant?

File: Modules/_ctypes/_ctypes.c [hunk 2]
```diff
@@ -4870,11 +4867,8 @@ PyCArrayType_from_ctype(PyObject *itemtype, Py_ssize_t length)
...
-    if (_PyDict_GetItemProxy(cache, key, &result) < 0) {
-        Py_DECREF(key);
-        return NULL;
-    }
-    if (result) {
+    if (_PyDict_GetItemProxy(cache, key, &result) != 0) {
+        // error or found
         Py_DECREF(key);
         return result;
     }
```
Comment (vstinner) [inline Modules/_ctypes/_ctypes.c]: Please add ``// not found`` after to be extra explicit, since this code pattern is not common.

**Review by vstinner — COMMENTED** [review]
Reviewing code with 3 cases (error, not found, found) is really hard to me, especially since you also change ownership and reference counting in the same PR. I cannot review such giant single PR. If you want me to review it, please create smaller PRs. For example, only change around 10 PyDict_Get call per PR. Just `Modules/_ctypes/_ctypes.c` alone is already a big piece.

File: Modules/_ctypes/_ctypes.c (hunk 1 above)
Comment (serhiy-storchaka) [reply Modules/_ctypes/_ctypes.c]: Yes, redundant. And the code here can be simplified even more.

File: Modules/_ctypes/_ctypes.c (hunk 2 above)
Comment (serhiy-storchaka) [reply Modules/_ctypes/_ctypes.c]: Added here and in other places.

**Comment by vstinner** [comment]
What's the status of that PR? Was it rebased on the main branch to retrieve other already merged changes? It's marked as a draft. I prefer to wait until it's not a draft to review it.

**Comment by serhiy-storchaka** [comment]
It was merged with `main` to include already merged changes. I extracted also Python/bytecodes.c in a separate PR #111978.

The remaining is relatively small and simple.

**Comment by serhiy-storchaka** [comment]
Split on more PRs.

**Comment by vstinner** [comment]
> Split on more PRs.

Is this PR ready for a review? It's marked as a draft and now has a conflict.

**Comment by serhiy-storchaka** [comment]
I left the combined PR to see the general image, but most of changes were already merged in other PRs.

**Pull request closed without merging**